cql3/statements/role-management-statements.cc

```diff
@@ -179,6 +180,8 @@ create_role_statement::execute(service::storage_proxy&,
             return grant_permissions_to_creator(cs);
         }).then([&state] () mutable {
             return state.get_client_state().update_per_role_params();
+        }).then([&state] () mutable {
+            return db::system_keyspace::update_per_session_params(state.get_client_state());
```

Comment:
ALTER ROLE may affect a different role than the one we are running.

It should be left to the code that updates the timeouts already connected sessions.

Reply:
Yeah, it's pretty much lazy person's placeholder for easy testing. Ultimately, all sessions should simply get updated after these settings change.

Reply:
OTOH, we could say the same thing about timeouts that we already say about permissions: the new settings only take hold after you log out and back in.

Reply:
@psarna it also doesn't affect other nodes. Let's drop  this and do it correctly later.

Reply:
> OTOH, we could say the same thing about timeouts that we already say about permissions: the new settings only take hold after you log out and back in.

That is actually pretty bad, since it means if you drop some permissions to restrict an application from doing bad things, you might not know you need to restart it.

Looks like we'll need a trigger mechanism to tell permissions and session watchers that something changed. I suppose the replica layer can broadcast a role_change event and cause caches to be flushed.

Reply:
Dropping it also makes the whole test suite invalid though :( So dropping this one is only feasible if we have some other mechanism for propagating the timeouts, because I'd rather not push this change without tests.

I can try adding proper propagation in this series as well, but it might end up being quite intrusive (literally, because chaining all client states in an intrusive list would be a way to go). Propagating to other shards would be quite easy, but as for other nodes, I guess we should observe schema changes and trigger an update when roles change?

Reply:
Alright, I thought about it more, and my decision is that I'll ensure proper propagation **within a node** in this series. It will enable testing and be a good first step. Then, we can perhaps touch inter-node propagation later?

Reply:
Sigh, `connection` class is already linked into an intrusive list and it makes it really easy to iterate over a snapshot of that list and update all the client states... the problem is that this intrusive list is from entirely another layer (it's a member of `cql_server`) and there's no way to reach it from role-updating statements. I would really prefer to avoid creating another intrusive list (over client states), but I don't see any clean alternatives.

Reply:
Maybe I'll just add a listener for a new `role_change` event.

Reply:
@psarna a role change event is not enough, how do you trigger it on other nodes?

Reply:
In a follow-up, naturally! (: But, back to being serious, I plan to implement local propagation first and leave inter-node for later, because this pull request keeps growing. Propagating locally is not perfect, but it enables majority of the test scenarios. And, after such an event is implemented, propagating to other nodes will also get easier, since it would all be about generating an event in a proper place.

Reply:
> > OTOH, we could say the same thing about timeouts that we already say about permissions: the new settings only take hold after you log out and back in.
> 
> That is actually pretty bad, since it means if you drop some permissions to restrict an application from doing bad things, you might not know you need to restart it.

I was talking more about role membership, which affects permissions indirectly.  I'm pretty sure that's not immediately propagated to all live sessions.  That's not ideal, but I doubt it's a surprise to any seasoned admin.

> Looks like we'll need a trigger mechanism to tell permissions and session watchers that something changed. I suppose the replica layer can broadcast a role_change event and cause caches to be flushed.

That works when roles are stored in Scylla but won't work when they're defined in an external source like LDAP.

Reply:
> I was talking more about role membership, which affects permissions indirectly. I'm pretty sure that's not immediately propagated to all live sessions.

I should clarify this: the changes to `system_auth.roles` is propagated to all nodes fairly soon, but that table is not queried afresh on every statement executed in a session.  It's queried at login, and the result is cached for some time.